### PR TITLE
Feature/communication calling server preview

### PR DESCRIPTION
**Pull request by moirf**
## Purpose

* ...

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: GetRelayConfiguration/Program.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+using Microsoft.MixedReality.WebRTC;
+
+namespace GetRelayConfiguration
+{
+    class Program
+    {
+        private static CommunicationRelayClient _communicationRelayClient;
+        private static CommunicationIdentityClient _communicationIdentityClient;
+
+        static async Task Main(string[] args)
```
Comment (richardcho-msft) [inline GetRelayConfiguration/Program.cs:19]: What does this do?

File: IncomingCallRouting/IncomingCallRouting/Utils/CallConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        /// <summary>
+        /// The publicly available participant id to transfer the incoming call.
+        /// </summary>
+        public string targetParticipant { get; private set; }
+
+        /// <summary>
+        /// The publicly available participant id to transfer the incoming call.
+        /// </summary>
+        public string ivrParticipant { get; private set; }
```
Comment (moirf) [inline IncomingCallRouting/IncomingCallRouting/Utils/CallConfiguration.cs:68]: @v-mausha update property as pascal casing - targetParticipant and ivrParticipant..

File: GetRelayConfiguration/Program.cs (hunk 1 above)
Comment (moirf) [reply GetRelayConfiguration/Program.cs:19]: This is not for incoming routing, but its changes are due to taking the changes from main branch in sync. This was written for getting underline servers.

File: IncomingCallRouting/IncomingCallRouting/Controllers/IncomingCallController.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    }
+                }
+                else if (cloudEvent.EventType.Equals("Microsoft.Communication.IncomingCall"))
+                {
+                    //Fetch incoming call context and ivr participant from request
+                    var eventData = request.ToString();
+                    if (eventData != null)
+                    {
+                        string incomingCallContext = eventData.Split("\"incomingCallContext\":\"")[1].Split("\"}")[0];
+                        string ivrParticipnat = eventData.Split("\"to\":{\"rawId\":\"")[1].Split("\",\"")[0];
```
Comment (fangchen0601) [inline IncomingCallRouting/IncomingCallRouting/Controllers/IncomingCallController.cs:60]: typo on ivrParticipnat

File: OutboundCallReminder/OutboundCallReminder/OutboundCallReminder.cs [hunk 4]
```diff
@@ -103,8 +103,8 @@ private async Task<CallConnection> CreateCallAsync(string targetPhoneNumber)
                 var target = new PhoneNumberIdentifier(targetPhoneNumber);
                 var createCallOption = new CreateCallOptions(
                     new Uri(callConfiguration.AppCallbackUrl),
-                    new List<CallMediaType> { CallMediaType.Audio },
-                    new List<CallingEventSubscriptionType> { CallingEventSubscriptionType.ParticipantsUpdated, CallingEventSubscriptionType.ToneReceived }
+                    new List<MediaType> { MediaType.Audio },
```
Comment (fangchen0601) [inline OutboundCallReminder/OutboundCallReminder/OutboundCallReminder.cs:106]: this should be CallMediaType. why this PR makes change to outbound call sample?

File: OutboundCallReminder/OutboundCallReminder/OutboundCallReminder.cs [hunk 5]
```diff
@@ -145,16 +145,17 @@ private async Task PlayAudioAsync()
...
                 };
 
                 Logger.LogMessage(Logger.MessageType.INFORMATION, "Performing PlayAudio operation");
-                var response = await callConnection.PlayAudioAsync(new Uri(callConfiguration.AudioFileUrl), playAudioRequest, reportCancellationToken).ConfigureAwait(false);
+                var response = await callConnection.PlayAudioAsync(playAudioRequest, reportCancellationToken).ConfigureAwait(false);
 
                 Logger.LogMessage(Logger.MessageType.INFORMATION, $"PlayAudioAsync response --> {response.GetRawResponse()}, Id: {response.Value.OperationId}, Status: {response.Value.Status}, OperationContext: {response.Value.OperationContext}, ResultInfo: {response.Value.ResultInfo}");
 
-                if (response.Value.Status == CallingOperationStatus.Running)
+                if (response.Value.Status == OperationStatus.Running)
```
Comment (fangchen0601) [inline OutboundCallReminder/OutboundCallReminder/OutboundCallReminder.cs:158]: seems this PR will revert back the latest changes on outbound call. could you double check?

File: IncomingCallRouting/IncomingCallRouting/Utils/IncomingCallHandler.cs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+                if (identifierKind == CommunicationIdentifierKind.UserIdentity)
+                {
+                    var response = await callConnection.TransferToParticipantAsync(new CommunicationUserIdentifier(targetParticipant), null, null, operationContext).ConfigureAwait(false);
+                    Logger.LogMessage(Logger.MessageType.INFORMATION, $"TransferParticipantAsync response --> {response.GetRawResponse()}, status: {response.Value.Status}  " +
+                        $"OperationContext: {response.Value.OperationContext}, OperationId: {response.Value.OperationId}, ResultDetails: {response.Value.ResultDetails}");
+                }
+                else if (identifierKind == CommunicationIdentifierKind.PhoneIdentity)
+                {
+                    var response = await callConnection.TransferToParticipantAsync(new PhoneNumberIdentifier(targetParticipant), null, null, operationContext).ConfigureAwait(false);
```
Comment (fangchen0601) [inline IncomingCallRouting/IncomingCallRouting/Utils/IncomingCallHandler.cs:304]: dupe code here when calling PMA, the TransferToParticipantAsync() takes CommunicationIdentifier as the argument. so you don't need to create separate CommunicationUserIdentifier and PhoneNumberIdentifier.

File: IncomingCallRouting/IncomingCallRouting/Utils/IncomingCallHandler.cs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+            var transferToParticipantCompleted = await transferToParticipantCompleteTask.Task.ConfigureAwait(false);
+            return transferToParticipantCompleted;
+        }
+
+        private void RegisterToTransferParticipantsResultEvent(string operationContext)
+        {
+            var transferToParticipantReceivedEvent = new NotificationCallback(async (CallingServerEventBase callEvent) =>
+            {
+                var transferParticipantUpdatedEvent = (ParticipantsUpdatedEvent)callEvent;
```
Comment (fangchen0601) [inline IncomingCallRouting/IncomingCallRouting/Utils/IncomingCallHandler.cs:317]: it should be TransferCallResultEvent

**Review by fangchen0601 — CHANGES_REQUESTED** [review]

File: OutboundCallReminder/OutboundCallReminder/OutboundCallReminder.cs (hunk 5 above)
Comment (v-mausha) [reply OutboundCallReminder/OutboundCallReminder/OutboundCallReminder.cs:158]: we got these OutboundCallReminder changes after fetching the latest changes from **azure-samples:main** branch.
Also these changes are working with latest beta2 and beta3 packages

File: IncomingCallRouting/IncomingCallRouting/EventHandler/EventDispatcher.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            else if (callEventBase is PlayAudioResultEvent)
+            {
+                var operationContext = ((PlayAudioResultEvent)callEventBase).OperationContext;
+                return BuildEventKey(CallingServerEventType.PlayAudioResultEvent.ToString(), operationContext);
+            }
+            else if (callEventBase is ParticipantsUpdatedEvent)
+            {
+                var callLegId = ((ParticipantsUpdatedEvent)callEventBase).CallConnectionId;
+                return BuildEventKey(CallingServerEventType.ParticipantsUpdatedEvent.ToString(), callLegId);
+            }
```
[the remark this answers is not in the document]
Comment (fangchen0601) [reply IncomingCallRouting/IncomingCallRouting/EventHandler/EventDispatcher.cs:95]: Then the C# SDK should be updated, it is in the swagger: https://github.com/Azure/azure-rest-api-specs/blob/acs-callingServer/2021-11-15-preview/specification/communication/data-plane/CallingServer/preview/2021-11-15-preview/communicationservicescallingserver.json

**Pull request closed without merging**